aten/src/ATen/VmapTransforms.cpp

```diff
@@ -94,33 +94,112 @@ static BatchDims computeFrontBatchDimsFromLevels(std::bitset<kVmapNumLevels> lev
+
+  // align the example dims (non-bdims dims) first
+  // aligned_sizes[-tensor_example_dim:] = tensor_sizes[-tensor_example_dim:]
+  std::copy(
+      physical_sizes.rbegin(),
+      physical_sizes.rbegin() + tensor_example_dim,
+      aligned_sizes.rbegin());
+
+  // align the bdims
+  int64_t level = 0;
+  int64_t dim = 0;
+  int64_t tensor_dim = 0;
+  for (int64_t level = 0; level < kVmapNumLevels; level++) {
+    if (!requested_levels[level]) {
+      continue;
+    }
+    if (tensor_levels[level]) {
+      aligned_sizes[dim] = physical_sizes[tensor_dim++];
+    }
+    dim++;
   }
-  for (int64_t dim = 0; dim < num_batch_dims; dim++) {
-    if (physical_sizes[dim] != reference_sizes[dim]) {
-      return false;
+
+  return physical_tensor.view(aligned_sizes);
+}
+
+static std::pair<std::bitset<kVmapNumLevels>,int64_t>
+getLevelsAndLargestExampleDim(TensorList logical_tensors) {
+  TORCH_INTERNAL_ASSERT(logical_tensors.size() > 0);
+  std::bitset<kVmapNumLevels> levels;
+  int64_t largest_example_dim = -1;
+  for (const auto& tensor : logical_tensors) {
+    auto* batched = maybeGetBatched(tensor);
+    if (batched) {
+      levels = levels | createLevelsBitset(batched->bdims());
+    }
+    auto ndim = tensor.dim();
+    if (ndim > largest_example_dim) {
+      largest_example_dim = ndim;
     }
   }
-  return true;
+  return { levels, largest_example_dim };
 }
 
-Tensor VmapPhysicalView::newLogicalFromPhysical(const Tensor& physical) {
+std::vector<VmapPhysicalView> BroadcastingVmapTransform::logicalToPhysical(TensorList logical_tensors) {
   TORCH_INTERNAL_ASSERT(
-      batchDimsAreSameSize(physical, tensor_, numBatchDims()),
-      "VmapPhysicalView::newLogicalFromPhysical(physical): expected batch dims ",
-      "of `physical` to be the same size as the batch dims from this VmapPhysicalView.");
-  return makeBatched(physical, computeFrontBatchDimsFromLevels(levels_));
-}
+      logical_tensors.size() == 2,
+      "This function has only been tested for two tensors. Please add more tests ",
+      "before removing this check ");
+
+  std::vector<VmapPhysicalView> result;
 
+  std::bitset<kVmapNumLevels> levels;
+  int64_t largest_example_dim;
+  std::tie(levels, largest_example_dim) = getLevelsAndLargestExampleDim(logical_tensors);
+
+  for (const auto& tensor : logical_tensors) {
+    // NB: It's possible that we didn't actually need to align `tensor`.
+    // For example, when adding two tensors of size (B, 2), and (3, 2), where
+    // the first Tensor is a BatchedTensor with batch dim B and the second is
+    // a regular Tensor, we will return views of size (B, 1, 2) and (1, 3, 2).
+    // However, the view on the second tensor is unnecessary: broadcasting
+    // semantics allow for the addition of two tensors of size (B, 1, 2) and (3, 2)!
+    //
+    // If this unnecessary view is a problem, consider optimizing it away in
+    // the future. This may involve creating a new type of VmapPhysicalView
```

Comment:
This comment here also suggests that perhaps BroadcastingVmapTransform should return a different kind of VmapPhysicalView. Right now, `VmapPhysicalView` must hold a `levels` bitset to perform the physical->logical map, but that `levels` bitset is also used to map logical (dims, shapes, etc) to physical (dims, shapes etc).

An alternative might be to separate the VmapPhysicalView::newLogicalFromPhysical from VmapPhysicalView

Reply:
Hot take: this optimization doesn't seem worth it.